src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs

```diff
@@ -2991,7 +2991,7 @@ private TypeSymbol StackMergeType(BoundExpression expr)
                     var conversionKind = conversion.ConversionKind;
                     if (conversionKind.IsImplicitConversion() &&
                         conversionKind != ConversionKind.MethodGroup &&
-                        conversionKind != ConversionKind.NullLiteral)
+                        conversionKind != ConversionKind.DefaultOrNullLiteral)
```

Comment:
I didn't manage to reach this with `null` or `default` either. The logic above in `StackMergeType` exits early except for array, interface or delegate types. But the local rewriter strips the bound conversion on an array, interface and delegate types (because they have a constant value).